File: windows/DisplayShot/Annotations/AnnotationRenderer.cs
using System.Globalization;
using System.Windows;
using System.Windows.Media;

namespace DisplayShot.Annotations;

/// <summary>Draws annotations into a DrawingContext. Units are canvas pixels; the caller sets up
/// any transform. Used by both the live overlay and the exporter.</summary>
public static class AnnotationRenderer
{
    public const double MarkerAlpha = 0.35;
    private static readonly Typeface TextTypeface = new(new FontFamily("Segoe UI"), FontStyles.Normal, FontWeights.SemiBold, FontStretches.Normal);

    /// <summary>Draws everything except redactions (pixel operations handled by ImageComposer).
    /// Markers are grouped so overlapping strokes never double-darken.</summary>
    public static void Draw(DrawingContext dc, IReadOnlyList<Annotation> items, Rect? clip, bool includeMarkers = true)
    {
        if (clip is { } c) dc.PushClip(new RectangleGeometry(c));

        if (includeMarkers && items.Any(a => a.IsMarker))
        {
            dc.PushOpacity(MarkerAlpha);
            foreach (var a in items)
                if (a is MarkerAnnotation m) DrawStroke(dc, m.Points, m.Stroke);
            dc.Pop();
        }

        foreach (var a in items)
        {
            if (a.IsMarker || a.IsRedaction) continue;
            DrawOne(dc, a);
        }

        if (clip is not null) dc.Pop();
    }

    /// <summary>Draws only the marker strokes, fully opaque (used for the multiply pass on export).</summary>
    public static void DrawMarkersOpaque(DrawingContext dc, IReadOnlyList<Annotation> items)
    {
        foreach (var a in items)
            if (a is MarkerAnnotation m) DrawStroke(dc, m.Points, m.Stroke);
    }

    /// <summary>Draws one annotation (used for the in-progress shape).</summary>
    public static void DrawOne(DrawingContext dc, Annotation a)
    {
        switch (a)
        {
            case PenAnnotation pen:
                DrawStroke(dc, pen.Points, pen.Stroke);
                break;
            case MarkerAnnotation marker:
                dc.PushOpacity(MarkerAlpha);
                DrawStroke(dc, marker.Points, marker.Stroke);
                dc.Pop();
                break;
            case LineAnnotation line:
                dc.DrawLine(MakePen(line.Stroke), line.From, line.To);
                break;
            case ArrowAnnotation arrow:
                DrawArrow(dc, arrow.From, arrow.To, arrow.Stroke);
                break;
            case RectangleAnnotation rect:
                dc.DrawRectangle(null, MakePen(rect.Stroke), rect.Rect);
                break;
            case TextAnnotation text:
                DrawText(dc, text);
                break;
        }
    }

    public static Pen MakePen(Stroke s)
    {
        var pen = new Pen(new SolidColorBrush(s.Color), s.Width)
        {
            StartLineCap = PenLineCap.Round,
            EndLineCap = PenLineCap.Round,
            LineJoin = PenLineJoin.Round,
        };
        pen.Freeze();
        return pen;
    }

    /// <summary>Smooth freehand stroke: quadratic curves through the midpoints.</summary>
    public static void DrawStroke(DrawingContext dc, IReadOnlyList<Point> pts, Stroke s)
    {
        if (pts.Count == 0) return;
        var pen = MakePen(s);
        if (pts.Count <= 2)
        {
            dc.DrawLine(pen, pts[0], pts[^1]);
            return;
        }
        var geometry = new StreamGeometry();
        using (var ctx = geometry.Open())
        {
            ctx.BeginFigure(pts[0], false, false);
            for (var i = 1; i < pts.Count - 1; i++)
            {
                var mid = new Point((pts[i].X + pts[i + 1].X) / 2, (pts[i].Y + pts[i + 1].Y) / 2);
                ctx.QuadraticBezierTo(pts[i], mid, true, false);
            }
            ctx.LineTo(pts[^1], true, false);
        }
        geometry.Freeze();
        dc.DrawGeometry(null, pen, geometry);
    }

    public static void DrawArrow(DrawingContext dc, Point p, Point q, Stroke s)
    {
        var dx = q.X - p.X;
        var dy = q.Y - p.Y;
        var len = Math.Sqrt(dx * dx + dy * dy);
        if (len < 0.5) return;
        var ux = dx / len;
        var uy = dy / len;
        var headLength = Math.Max(12, s.Width * 4);
        var headWidth = Math.Max(10, s.Width * 3.2);
        var basePoint = new Point(q.X - ux * headLength, q.Y - uy * headLength);
        var shaftEnd = new Point(q.X - ux * headLength * 0.7, q.Y - uy * headLength * 0.7);

        dc.DrawLine(MakePen(s), p, shaftEnd);

        var nx = -uy;
        var ny = ux;
        var left = new Point(basePoint.X + nx * headWidth / 2, basePoint.Y + ny * headWidth / 2);
        var right = new Point(basePoint.X - nx * headWidth / 2, basePoint.Y - ny * headWidth / 2);
        var head = new StreamGeometry();
        using (var ctx = head.Open())
        {
            ctx.BeginFigure(q, true, true);
            ctx.LineTo(left, false, false);
            ctx.LineTo(right, false, false);
        }
        head.Freeze();
        dc.DrawGeometry(new SolidColorBrush(s.Color), null, head);
    }

    public static FormattedText MakeFormattedText(TextAnnotation t, Brush brush) =>
        new(t.Text, CultureInfo.CurrentUICulture, FlowDirection.LeftToRight, TextTypeface, t.FontSize, brush, 1.0);

    public static Size MeasureText(TextAnnotation t)
    {
        var ft = MakeFormattedText(t, Brushes.Black);
        return new Size(Math.Ceiling(ft.WidthIncludingTrailingWhitespace), Math.Ceiling(ft.Height));
    }

    /// <summary>Text with a soft dark shadow so it stays legible on any background.</summary>
    public static void DrawText(DrawingContext dc, TextAnnotation t)
    {
        var shadow = MakeFormattedText(t, new SolidColorBrush(Color.FromArgb(0x8C, 0, 0, 0)));
        var offset = Math.Max(1, t.FontSize / 18);
        dc.DrawText(shadow, new Point(t.Origin.X + offset, t.Origin.Y + offset));
        dc.DrawText(MakeFormattedText(t, new SolidColorBrush(t.Color)), t.Origin);
    }
}

File: macos/Sources/DisplayShotKit/Annotations/AnnotationRenderer.swift
import AppKit

/// Draws annotations into a y-down CGContext whose units are points.
/// The caller must have `NSGraphicsContext.current` set (flipped) so text can be drawn.
enum AnnotationRenderer {
    static let markerAlpha: CGFloat = 0.35

    static func draw(_ items: [Annotation], in ctx: CGContext, clip: CGRect?) {
        ctx.saveGState()
        if let clip { ctx.clip(to: clip) }

        // All marker strokes are drawn opaque into one transparency layer that is composited
        // once with alpha + multiply, so overlapping strokes never darken twice.
        let markers = items.filter { $0.isMarker }
        if !markers.isEmpty {
            ctx.saveGState()
            ctx.setAlpha(markerAlpha)
            ctx.setBlendMode(.multiply)
            ctx.beginTransparencyLayer(auxiliaryInfo: nil)
            for m in markers {
                if case .marker(let pts, let s) = m.kind { strokePath(pts, s, in: ctx) }
            }
            ctx.endTransparencyLayer()
            ctx.restoreGState()
        }

        for a in items where !a.isMarker && !a.isRedaction {
            draw(a, in: ctx)
        }
        ctx.restoreGState()
    }

    /// Draws one annotation (used for the in-progress shape). Redactions are pixel operations
    /// handled by `Redactor`, so they are skipped here.
    static func draw(_ a: Annotation, in ctx: CGContext) {
        switch a.kind {
        case .pen(let pts, let s):
            strokePath(pts, s, in: ctx)
        case .marker(let pts, let s):
            ctx.saveGState()
            ctx.setAlpha(markerAlpha)
            ctx.setBlendMode(.multiply)
            strokePath(pts, s, in: ctx)
            ctx.restoreGState()
        case .line(let p, let q, let s):
            applyStroke(s, ctx)
            ctx.move(to: p)
            ctx.addLine(to: q)
            ctx.strokePath()
        case .arrow(let p, let q, let s):
            drawArrow(from: p, to: q, s, in: ctx)
        case .rectangle(let r, let s):
            applyStroke(s, ctx)
            ctx.stroke(r)
        case .text(let t):
            drawText(t, in: ctx)
        case .redact:
            break
        }
    }

    static func applyStroke(_ s: Stroke, _ ctx: CGContext) {
        ctx.setStrokeColor(s.color)
        ctx.setLineWidth(s.width)
        ctx.setLineCap(.round)
        ctx.setLineJoin(.round)
    }

    /// Smooth freehand path through the points (quadratic curves through midpoints).
    static func strokePath(_ pts: [CGPoint], _ s: Stroke, in ctx: CGContext) {
        guard let first = pts.first else { return }
        applyStroke(s, ctx)
        let path = CGMutablePath()
        path.move(to: first)
        if pts.count <= 2 {
            path.addLine(to: pts.last ?? first)
        } else {
            for i in 1..<(pts.count - 1) {
                let mid = CGPoint(x: (pts[i].x + pts[i + 1].x) / 2, y: (pts[i].y + pts[i + 1].y) / 2)
                path.addQuadCurve(to: mid, control: pts[i])
            }
            path.addLine(to: pts[pts.count - 1])
        }
        ctx.addPath(path)
        ctx.strokePath()
    }

    static func drawArrow(from p: CGPoint, to q: CGPoint, _ s: Stroke, in ctx: CGContext) {
        let dx = q.x - p.x, dy = q.y - p.y
        let len = hypot(dx, dy)
        guard len > 0.5 else { return }
        let ux = dx / len, uy = dy / len
        let headLength = max(12, s.width * 4)
        let headWidth = max(10, s.width * 3.2)
        let base = CGPoint(x: q.x - ux * headLength, y: q.y - uy * headLength)
        let shaftEnd = CGPoint(x: q.x - ux * headLength * 0.7, y: q.y - uy * headLength * 0.7)

        applyStroke(s, ctx)
        ctx.move(to: p)
        ctx.addLine(to: shaftEnd)
        ctx.strokePath()

        let nx = -uy, ny = ux
        let left = CGPoint(x: base.x + nx * headWidth / 2, y: base.y + ny * headWidth / 2)
        let right = CGPoint(x: base.x - nx * headWidth / 2, y: base.y - ny * headWidth / 2)
        ctx.setFillColor(s.color)
        ctx.move(to: q)
        ctx.addLine(to: left)
        ctx.addLine(to: right)
        ctx.closePath()
        ctx.fillPath()
    }

    static func textAttributes(_ t: TextAnnotation) -> [NSAttributedString.Key: Any] {
        let shadow = NSShadow()
        shadow.shadowColor = NSColor(white: 0, alpha: 0.55)
        shadow.shadowBlurRadius = 2
        shadow.shadowOffset = NSSize(width: 0, height: -1)
        return [
            .font: NSFont.systemFont(ofSize: t.fontSize, weight: .semibold),
            .foregroundColor: NSColor(cgColor: t.color) ?? .white,
            .shadow: shadow,
        ]
    }

    static func textSize(_ t: TextAnnotation) -> CGSize {
        NSAttributedString(string: t.string, attributes: textAttributes(t)).size()
    }

    static func drawText(_ t: TextAnnotation, in ctx: CGContext) {
        let str = NSAttributedString(string: t.string, attributes: textAttributes(t))
        str.draw(at: t.origin)
    }
}
